web/viewer.js

```diff
@@ -590,6 +586,112 @@ var PDFViewerApplication = {
     ).then(null, downloadByUrl);
   },
 
+
+  requestPrint: function pdfViewRequestPrint() {
+//#if !(CHROME || GENERIC) || !WITH_EDITING
+//  window.print();
+//#else
+    var printAction = {
+      embed: null,
+      pdfBlobUrl: '',
+      handleEvent: function onWindowFocus() {
+        disposePrintAction(printAction);
+      },
+    };
+    if (this._printAction) {
+      if (!this._printAction.embed) {
+        // Do not restart a print request if the previous one is still pending.
+        return;
+      }
+      disposePrintAction(this._printAction);
+    }
+    this._printAction = printAction;
+
+    canUsePDFium(function yes() {
+      PDFViewerApplication.pdfDocument.getDataForPrinting().then(function(res) {
+        printAction.pdfBlobUrl = res.pdfBlobUrl;
+        printAction.embed = createPDFiumEmbed();
+        printAction.embed.src = printAction.pdfBlobUrl;
+        document.body.appendChild(printAction.embed);
+        // Clean up when the print dialog is closed.
+        window.addEventListener('focus', printAction);
+      }, onFallback);
+    }, onFallback);
+
+    function onFallback() {
+      disposePrintAction(printAction);
+      window.print();
+    }
+
+    function disposePrintAction(printAction) {
+      window.removeEventListener('focus', printAction);
+      if (printAction.embed) {
+        printAction.embed.remove();
+      }
+      if (printAction.pdfBlobUrl) {
+        URL.revokeObjectURL(printAction.pdfBlobUrl);
+      }
+      if (PDFViewerApplication._printAction === printAction) {
+        PDFViewerApplication._printAction = null;
+      }
+    }
+
+    // Create an off-screen <embed> element.
+    function createPDFiumEmbed() {
+      var embed = document.createElement('embed');
+      embed.type = 'application/x-google-chrome-pdf';
+      embed.style.position = 'fixed';
+      embed.style.width = '9px';
+      embed.style.height = '9px';
+      embed.style.top = '-9990px;';
+      embed.style.left = '-999px';
+      return embed;
+    }
+
+    // Check whether Chromium's built-in PDF renderer (PDFium) plugin is enabled
+    // and usable without click-to-play. The result is not cached because the
+    // user can change the value via the browser settings.
+    // This feature detection never has false positives, only false negatives.
+    function canUsePDFium(yes, no) {
```

Comment:
`canUsePDFium` should return a promise and resolve/reject it instead of yes/no callbacks.